### PR TITLE
Add RetargetModifier3D for realtime retarget to keep original rest

### LINKED ISSUE #3379: Implement Skeleton re-targeting node
### Describe the project you are working on

Godot

### Describe the problem or limitation you are having in your project

The only way to re-use animations is use compatible skeletons with similarly named bones and proportions. This can be significantly limiting for sharing animations or using animations from common sources such as Mixamo.

While in-engine re-targeting is generally not a technique advised for production (as it makes the amount of bones you can use to animate rather limiting, thus makes rigging looks worse, for reference most studios do this within the 3D DCC together with IK), there are still cases where it makes a lot of sense, such as body tracking, prototyping, or simply Indie development where creating custom animations for models can be expensive.

Note, this is an actual implementation proposal based on #2619, so it supersedes it.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

The proposal here is to implement re-targeting as a special node. Mostly to make Skeleton less bloated and to reduce the need to make children scenes editable in order to configure them.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

A new **SkeletonRetarget** node will be provided by the engine. This node will have the following properties:

* source : **NodePath**
* target : **NodePath**
* retargeter : **Retargeter** (resource)

**Retargeter** is a resource that does the retargeting logic. It is a virtual abstract class that does no logic, but it can be inherited to implement a retargeter. Godot will provide a default implementation: **HumanoidRetargeter** which will come with a default humanoid skeleton. 

Editing **HumanoidRetargeter** will be similar to Unity/Maya/etc., except that you must set the source bone and destination bone. 

Given the **Retargeter** is a resource, it will be possible to save it to disk to use it in other scenes.

Eventually, it should be possible to provide other more precise retargeters with more complex models via GDExtension if desired.

### If this enhancement will not be used often, can it be worked around with a few lines of script?

Animation is core

### Is there a reason why this should be core and not an add-on in the asset library?

Animation is core

### PR DESCRIPTION
**Pull request by TokageItLab**
Closes https://github.com/godotengine/godot-proposals/issues/3379

## RetargetModifier3D

Add RetargetModifier3D. Currently Godot performs retargeting by overwriting rest. However, that means discarding the bone rest which set in the DCC, and it is quite possible that the axis of rest will be in an unintended direction. It makes re-targeting useless for specifying IK pole vectors and setting constraints for specific axes.

This PR solves this problem by retargeting the parent skeleton with the rest defined in the SkeletonProfile and the child skeleton holding the original rest with the RetargetModifier that will be added in this PR. Also keep in mind that parenting is required to guarantee process order.

image

The Profile specified to RetargetModifier3D is used only as a list of bone names, so if you want to retarget only specific bones, you can create a custom SkeletonProfile with arbitrary bone names only and set it to the RetargetModifier3D.

It has the following two types of retargets by `Use Global Pose` option.

### `Use Global Pose` option

If enabled, retargeting is enabled when the parent of the source bone is deformed, even if the target bone is less than the source bone. However, as a limitation, the lengths of the bones must be exactly equal. For example a specific use case, adding a dummy bone with zero length helps share animation between models with different numbers of bones.

https://github.com/user-attachments/assets/44a86290-4950-4d2b-9ebb-99036e7e0e85

If disabled, retargeting is not performed when the parent of the source bone is deformed if the target bone is less than the source bone. However, it is possible to retarget bones of different lengths without destroying the appearance of the bones. For example a specific use case, you can retarget a chibi character from a skeleton that matches your real body shape for XR motion tracking. In addition, position, rotation, and scale can be handled individually.

image

### Add option to separate unmapped bone library

image

### Add option to use RetargetModifier to keep original rest

image image

Note that the Rename Bone option must be enabled.

image

### CODE REVIEW

**Comment by fire** [comment]
Is there particular tests you want us to look at? I'll try to review this as soon as possible.

**Comment by TokageItLab** [comment]
[retarget-modifier-test.zip](https://github.com/user-attachments/files/17265445/retarget-modifier-test.zip)
---
Godette / CC by [SirRichard94](https://github.com/SirRichard94/low-poly-godette)
Blender Chan! / CC by [SearKitchen](https://sketchfab.com/searkitchen)
Alicia Solid / [DWANGO Co., Ltd.](https://3d.nicovideo.jp/alicia/)

**Comment by TokageItLab** [comment]
It may be possible to modify it so that Skeleton can create a children list even if it is not a direct child of RetargetModifier like:
```
ParentSkeleton
└RetargetModifier
　└Node
　　└ChildSkeleton <- detect
```
but it needs to be tested.

I will try to improve this in a later Follow up PR, as it could be applied to SkeletonModifier lists that are direct children of Skeleton to organize tree view like:
```
Skeleton
└Node
│├ModifierA_1
│└ModifierA_2
└Node_2
　├ModifierB_1
　└ModifierA_2
```

**Comment by fire** [comment]
I have tested this on https://github.com/V-Sekai/godot-humanoid-project where we retarget a character body without a UpperChest to a character body that has does not have a UpperChest but only has a Chest.

**Review by SaracenOne — APPROVED** [review]
Was discussed and demonstrated in the Animation meeting and seems fine 👍

File: scene/3d/retarget_modifier_3d.cpp [hunk 1]
```diff
@@ -0,0 +1,441 @@
...
+void RetargetModifier3D::_retarget_global_pose() {
+	Skeleton3D *source_skeleton = get_skeleton();
+	if (profile.is_null() || !source_skeleton) {
+		return;
+	}
+
+	LocalVector<Transform3D> source_poses;
+	if (influence < 1.0) {
+		for (int i = 0; i < source_bone_ids.size(); i++) {
+			source_poses.push_back(source_bone_ids[i] < 0 ? Transform3D() : source_skeleton->get_bone_global_rest(source_bone_ids[i]).interpolate_with(source_skeleton->get_bone_global_pose(source_bone_ids[i]), influence));
```
Comment (AThousandShips) [inline scene/3d/retarget_modifier_3d.cpp:279]:
```suggestion
		for (int source_bone_id : source_bone_ids) {
			source_poses.push_back(source_bone_id < 0 ? Transform3D() : source_skeleton->get_bone_global_rest(source_bone_id).interpolate_with(source_skeleton->get_bone_global_pose(source_bone_id), influence));
```

File: scene/3d/retarget_modifier_3d.cpp [hunk 2]
```diff
@@ -0,0 +1,441 @@
...
+		}
+	} else {
+		for (int i = 0; i < source_bone_ids.size(); i++) {
+			source_poses.push_back(source_bone_ids[i] < 0 ? Transform3D() : source_skeleton->get_bone_global_pose(source_bone_ids[i]));
```
Comment (AThousandShips) [inline scene/3d/retarget_modifier_3d.cpp:283]:
```suggestion
		for (int source_bone_id : source_bone_ids) {
			source_poses.push_back(source_bone_id < 0 ? Transform3D() : source_skeleton->get_bone_global_pose(source_bone_id));
```

File: scene/3d/retarget_modifier_3d.cpp [hunk 3]
```diff
@@ -0,0 +1,441 @@
...
+void RetargetModifier3D::_retarget_pose() {
+	Skeleton3D *source_skeleton = get_skeleton();
+	if (profile.is_null() || !source_skeleton) {
+		return;
+	}
+
+	LocalVector<Transform3D> source_poses;
+	if (influence < 1.0) {
+		for (int i = 0; i < source_bone_ids.size(); i++) {
+			source_poses.push_back(source_bone_ids[i] < 0 ? Transform3D() : source_skeleton->get_bone_rest(source_bone_ids[i]).interpolate_with(source_skeleton->get_bone_pose(source_bone_ids[i]), influence));
```
Comment (AThousandShips) [inline scene/3d/retarget_modifier_3d.cpp:313]:
```suggestion
		for (int source_bone_id : source_bone_ids) {
			source_poses.push_back(source_bone_id < 0 ? Transform3D() : source_skeleton->get_bone_rest(source_bone_id).interpolate_with(source_skeleton->get_bone_pose(source_bone_id), influence));
```

File: scene/3d/retarget_modifier_3d.cpp [hunk 4]
```diff
@@ -0,0 +1,441 @@
...
+		}
+	} else {
+		for (int i = 0; i < source_bone_ids.size(); i++) {
+			source_poses.push_back(source_bone_ids[i] < 0 ? Transform3D() : source_skeleton->get_bone_pose(source_bone_ids[i]));
```
Comment (AThousandShips) [inline scene/3d/retarget_modifier_3d.cpp:317]:
```suggestion
		for (int source_bone_id : source_bone_ids) {
			source_poses.push_back(source_bone_id < 0 ? Transform3D() : source_skeleton->get_bone_pose(source_bone_id));
```

File: scene/resources/skeleton_profile.cpp [hunk 5]
```diff
@@ -269,6 +269,14 @@ int SkeletonProfile::find_bone(const StringName &p_bone_name) const {
 	return -1;
 }
 
+PackedStringArray SkeletonProfile::get_bone_names() {
+	PackedStringArray s;
+	for (int i = 0; i < bones.size(); i++) {
+		s.push_back(bones[i].bone_name);
```
Comment (AThousandShips) [inline scene/resources/skeleton_profile.cpp:275]:
```suggestion
	for (const SkeletonProfileBone &bone : bones) {
		s.push_back(bone.bone_name);
```

File: editor/import/3d/post_import_plugin_skeleton_track_organizer.cpp [hunk 6]
```diff
@@ -114,11 +123,36 @@ void PostImportPluginSkeletonTrackOrganizer::internal_process(InternalImportCate
...
 					anim->remove_track(remove_indices[i]);
 				}
 			}
+
+			if (unmapped_al->get_animation_list_size() == 0) {
+				unmapped_al.unref();
+			} else {
+				if (separate_unmapped_bones == 2) {
+					ap->add_animation_library("unmapped_bones", unmapped_al);
+				}
```
Comment (AThousandShips) [inline editor/import/3d/post_import_plugin_skeleton_track_organizer.cpp:154]:
```suggestion
			} else if (separate_unmapped_bones == 2) {
				ap->add_animation_library("unmapped_bones", unmapped_al);
```

File: editor/import/3d/post_import_plugin_skeleton_track_organizer.cpp [hunk 7]
```diff
@@ -114,11 +123,36 @@ void PostImportPluginSkeletonTrackOrganizer::internal_process(InternalImportCate
...
 
+				if (separate_unmapped_bones == 2 && !unmapped_bone_indices.is_empty()) {
+					Ref<Animation> unmapped_anim = anim->duplicate();
+					Vector<int> to_delete;
+					to_delete.append_array(mapped_bone_indices);
+					to_delete.append_array(remove_indices);
+					to_delete.sort();
+					to_delete.reverse();
+					for (int i = 0; i < to_delete.size(); i++) {
+						unmapped_anim->remove_track(to_delete[i]);
```
Comment (AThousandShips) [inline editor/import/3d/post_import_plugin_skeleton_track_organizer.cpp:134]:
```suggestion
					for (int E : to_delete) {
						unmapped_anim->remove_track(E);
```

File: doc/classes/RetargetModifier3D.xml [hunk 8]
```diff
@@ -0,0 +1,34 @@
+<?xml version="1.0" encoding="UTF-8" ?>
+<class name="RetargetModifier3D" inherits="SkeletonModifier3D" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:noNamespaceSchemaLocation="../class.xsd">
+	<brief_description>
+		A modifier to transfer parent skeleton poses (or global poses) to child skeletons in model space with different rests.
+	</brief_description>
+	<description>
+		Retrieves the pose (or global pose) relative to the parent Skeleton's rest in model space and transfers it to the child Skeleton.
+		This modifier rewrites the pose of the child skeleton directly in the parent skeleton's update process. This means that it overwrites the mapped bone pose set in the normal process on the target skeleton. If you want to set the target skeleton bone pose after retargeting, you will need to add a child [SkeletonModifier3D] to the target skeleton and thereby modify the pose.
```
Comment (AThousandShips) [inline doc/classes/RetargetModifier3D.xml:8]:
```suggestion
		This modifier rewrites the pose of the child skeleton directly in the parent skeleton's update process. This means that it overwrites the mapped bone pose set in the normal process on the target skeleton. If you want to set the target skeleton bone pose after retargeting, you will need to add a [SkeletonModifier3D] child to the target skeleton and thereby modify the pose.
```

File: doc/classes/RetargetModifier3D.xml [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+		</member>
+		<member name="scale_enabled" type="bool" setter="set_scale_enabled" getter="is_scale_enabled" default="true">
+			If [code]true[/code], allows to retarget the scale.
+		</member>
+		<member name="use_global_pose" type="bool" setter="set_use_global_pose" getter="is_using_global_pose" default="false">
+			If [code]false[/code], in case the target skeleton has fewer bones than the source skeleton, the source bone parent's transform will be ignored.
+			Instead, it is possible to retarget between models with different body shapes, and position, rotation, and scale can be retargeted separately.
+			If [code]true[/code], Retargeting is performed taking into account global pose.
+			In case the target skeleton has fewer bones than the source skeleton, the source bone parent's transform is taken into account. However, bone length between skeletons must match exactly, if not, the bones will be forced to expand or shrink.
+			This is useful for using dummy bone which length is [b]0[/b] to match postures when retargeting between models with different number of bones.
```
Comment (AThousandShips) [inline doc/classes/RetargetModifier3D.xml:31]:
```suggestion
			This is useful for using dummy bone with length [code]0[/code] to match postures when retargeting between models with different number of bones.
```

**Review by AThousandShips — APPROVED** [review]

**Comment by fire** [comment]
@TokageItLab Does this resolve the `Add a bone expander (to be able to change the local scale)` priority?

**Comment by TokageItLab** [comment]
@fire No, it should be resolved with the option to add a cancel bone by constrain as discussed in https://github.com/godotengine/godot-proposals/discussions/9885#discussioncomment-9665995; It adds a constrained bone that applies the reverse scale to the child of the bone you want to cancel the scale. We have discussed this with lyuma in the past and we agree.

Canceling non-uniform scales requires allowing non-orthogonal scales in the bone pose, but non-orthogonal scales do not conform to the glTF specification and are incompatible with the TRS3D track, which prevents compression and hinders the implementation of streaming animation playback.

**Comment by Repiteo** [comment]
Thanks!

**Pull request merged**